src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,277 @@
+        }
+
+        public async Task<(MSB.Evaluation.Project project, DiagnosticLog log)> LoadProjectAsync(
+            string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        {
+            var log = new DiagnosticLog();
+
+            try
+            {
+                var project = FindProject(path, globalProperties, cancellationToken);
+                if (project != null)
+                {
+                    return (project, log);
+                }
+
+                using (var stream = FileUtilities.OpenAsyncRead(path))
+                using (var readStream = await SerializableBytes.CreateReadableStreamAsync(stream, cancellationToken).ConfigureAwait(false))
+                using (var xmlReader = XmlReader.Create(readStream, s_xmlReaderSettings))
+                {
+                    var xml = MSB.Construction.ProjectRootElement.Create(xmlReader, _projectCollection);
+
+                    // When constructing a project from an XmlReader, MSBuild cannot determine the project file path.  Setting the
+                    // path explicitly is necessary so that the reserved properties like $(MSBuildProjectDirectory) will work.
+                    xml.FullPath = path;
+
+                    project = new MSB.Evaluation.Project(xml, globalProperties, toolsVersion: null, _projectCollection);
+
+                    return (project, log);
+                }
+            }
+            catch (Exception e)
+            {
+                log.Add(e, path);
+                return (project: null, log);
+            }
+        }
+
+        public async Task<string> TryGetOutputFilePathAsync(
+            string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        {
+            var (project, _) = await LoadProjectAsync(path, globalProperties, cancellationToken).ConfigureAwait(false);
+            return project?.GetPropertyValue(PropertyNames.TargetPath);
+        }
+
+        public void Start()
+        {
+            if (_started)
+            {
+                throw new InvalidOperationException();
+            }
+
+            var buildParameters = new MSB.Execution.BuildParameters(_projectCollection);
+            buildParameters.Loggers = new MSB.Framework.ILogger[] { _logger };
+
+            MSB.Execution.BuildManager.DefaultBuildManager.BeginBuild(buildParameters);
+
+            _started = true;
+        }
+
+        public void Stop()
+        {
+            if (!_started)
+            {
+                throw new InvalidOperationException();
+            }
+
+            MSB.Execution.BuildManager.DefaultBuildManager.EndBuild();
+            _started = false;
+        }
+
+        public Task<MSB.Execution.ProjectInstance> BuildProjectAsync(
+            MSB.Evaluation.Project project, DiagnosticLog log, CancellationToken cancellationToken)
+        {
+            var targets = new[] { TargetNames.Compile, TargetNames.CoreCompile };
```

Comment:
I don't understand why we need to run `Compile` here.  VS only runs `CoreCompile` for design time builds.  I know you didn't add this, but I suspect it was added to hack around some other problem with xaml compiler invocation that we should probably clean up.

Reply:
Note that the sample I was given for design-time batch build (also called "Compile;CoreCompile"). I just dug that up and sent it to you as well.